### PR TITLE
Modals

### PR DESCRIPTION
**Pull request by FedorLap2006**
This PR contains implementation for newly introduced [modal interactions](https://devsnek.notion.site/devsnek/Modal-aka-Form-Interactions-e839b3dd8c214eb08f950764a8328e36) and text input component.

Useful links:

- [Docs](https://devsnek.notion.site/devsnek/Modal-aka-Form-Interactions-e839b3dd8c214eb08f950764a8328e36)
- https://github.com/discord/discord-api-docs/pull/4459

### CODE REVIEW

**Comment by FedorLap2006** [comment]
**Important note**: this PR will be **undrafted after the official release**

**Comment by Nv7-GitHub** [comment]
YESSS

**Comment by FedorLap2006** [comment]
Undrafting it due to https://github.com/discord/discord-api-docs/pull/4459.

File: components.go [hunk 1]
```diff
@@ -195,3 +198,42 @@ func (m SelectMenu) MarshalJSON() ([]byte, error) {
...
+		inputText
+		Type ComponentType `json:"type"`
+	}{
+		inputText: inputText(m),
+		Type:      m.Type(),
+	})
+}
+
+// TextStyleType is style of text in InputText component.
+type TextStyleType uint
```
Comment (TopiSenpai) [inline components.go:233]: shouldn't this be named `TextInputStyle`?

File: components.go [hunk 2]
```diff
@@ -195,3 +198,42 @@ func (m SelectMenu) MarshalJSON() ([]byte, error) {
...
+// InputText represents text input component.
+type InputText struct {
+	CustomID    string        `json:"custom_id,omitempty"`
+	Label       string        `json:"label"`
+	Style       TextStyleType `json:"style"`
+	Placeholder string        `json:"placeholder,omitempty"`
+	Value       string        `json:"value,omitempty"`
+	Required    bool          `json:"required"`
+	MinLength   int           `json:"min_length"`
+	MaxLength   int           `json:"max_length,omitempty"`
```
Comment (TopiSenpai) [inline components.go]:
```suggestion
type TextInput struct {
	CustomID    string        `json:"custom_id"`
	Label       string        `json:"label,omitempty"`
	Style       TextInputStyle `json:"style"`
	Placeholder string        `json:"placeholder,omitempty"`
	Value       string        `json:"value,omitempty"`
	Required    bool          `json:"required,omitempty"`
	MinLength   int           `json:"min_length,omitempty"`
	MaxLength   int           `json:"max_length,omitempty"`
```

File: components.go [hunk 3]
```diff
@@ -13,6 +13,7 @@ const (
 	ActionsRowComponent ComponentType = 1
 	ButtonComponent     ComponentType = 2
 	SelectMenuComponent ComponentType = 3
+	InputTextComponent  ComponentType = 4
```
Comment (TopiSenpai) [inline components.go]:
```suggestion
	TextInputComponent  ComponentType = 4
```

File: components.go [hunk 4]
```diff
@@ -42,6 +43,8 @@ func (umc *unmarshalableMessageComponent) UnmarshalJSON(src []byte) error {
 		umc.MessageComponent = &Button{}
 	case SelectMenuComponent:
 		umc.MessageComponent = &SelectMenu{}
+	case InputTextComponent:
+		umc.MessageComponent = &InputText{}
```
Comment (TopiSenpai) [inline components.go]:
```suggestion
	case TextInputComponent:
		umc.MessageComponent = &TextInput{}
```

**Review by TopiSenpai — CHANGES_REQUESTED** [review]

File: components.go [hunk 5]
```diff
@@ -195,3 +198,42 @@ func (m SelectMenu) MarshalJSON() ([]byte, error) {
...
+// TextInput represents text input component.
+type TextInput struct {
+	CustomID    string        `json:"custom_id,omitempty"`
+	Label       string        `json:"label"`
+	Style       TextStyleType `json:"style"`
+	Placeholder string        `json:"placeholder,omitempty"`
+	Value       string        `json:"value,omitempty"`
+	Required    bool          `json:"required"`
+	MinLength   int           `json:"min_length"`
+	MaxLength   int           `json:"max_length,omitempty"`
```
Comment (TopiSenpai) [inline components.go:211]:
```suggestion
type TextInput struct {
	CustomID    string        `json:"custom_id"`
	Label       string        `json:"label"`
	Style       TextInputStyle `json:"style"`
	Placeholder string        `json:"placeholder,omitempty"`
	Value       string        `json:"value,omitempty"`
	Required    bool          `json:"required,omitempty"`
	MinLength   int           `json:"min_length,omitempty"`
	MaxLength   int           `json:"max_length,omitempty"`
```

File: components.go [hunk 6]
```diff
@@ -195,3 +198,42 @@ func (m SelectMenu) MarshalJSON() ([]byte, error) {
...
+	})
+}
+
+// TextInputStyleType is style of text in TextInput component.
+type TextInputStyleType uint
+
+// Text styles
+const (
+	TextInputShort     TextStyleType = 1
+	TextInputParagraph TextStyleType = 2
```
Comment (TopiSenpai) [inline components.go:238]:
```suggestion
	TextInputStyleShort     TextStyleType = 1
	TextInputStyleParagraph TextStyleType = 2
```

File: components.go [hunk 7]
```diff
@@ -195,3 +198,42 @@ func (m SelectMenu) MarshalJSON() ([]byte, error) {
...
+		inputText
+		Type ComponentType `json:"type"`
+	}{
+		inputText: inputText(m),
+		Type:      m.Type(),
+	})
+}
+
+// TextInputStyleType is style of text in TextInput component.
+type TextInputStyleType uint
```
Comment (TopiSenpai) [inline components.go:233]:
```suggestion
type TextInputStyle uint
```

**Review by TopiSenpai — CHANGES_REQUESTED** [review]

File: components.go (hunk 6 above)
Comment (FedorLap2006) [reply components.go:238]: While it's a convention to name it `TextInputStyleShort`, `TextInputShort` looks better.

Comment (TopiSenpai) [reply components.go:238]: while it may sound better breaking conventions just hurts usability

**Comment by FedorLap2006** [comment]
Should be good now.

**Pull request merged**